crates/metrics/src/label.rs

```diff
@@ -0,0 +1,545 @@
+    enum_ordinalize::Ordinalize,
+    metrics::Label,
+    parking_lot::Mutex,
+    smallvec::SmallVec,
+    std::{borrow::Borrow, collections::HashMap, sync::Arc},
+};
+
+pub type DynamicLabels = SmallVec<[Label; 4]>;
+pub type StaticLabels = &'static [(&'static str, &'static str)];
+
+pub type Labeled<T, A> = WithLabel<A, T>;
+pub type Labeled2<T, A, B> = WithLabel<A, WithLabel<B, T>>;
+pub type Labeled3<T, A, B, C> = WithLabel<A, WithLabel<B, WithLabel<C, T>>>;
+pub type Labeled4<T, A, B, C, D> = WithLabel<A, WithLabel<B, WithLabel<C, WithLabel<D, T>>>>;
+
+pub trait DynamicLabel<M> {
+    type MetricCollection;
+}
+
+pub trait ResolveLabels<L> {
+    type Target;
+
+    fn resolve_labels(&self, labels: L) -> &Self::Target;
+}
+
+/// Adds dynamically resolved (in runtime) label to `M`.
+pub struct WithLabel<L: DynamicLabel<M>, M> {
+    collection: L::MetricCollection,
+}
+
+impl<L, M> WithLabel<L, M>
+where
+    L: DynamicLabel<M>,
+{
+    /// Resolves a single dynamic label and finds the underlying metric.
+    pub fn resolve_label<T>(&self, label: T) -> &M
+    where
+        Self: ResolveLabels<(T,), Target = M>,
+    {
+        self.resolve_labels((label,))
+    }
+
+    /// Resolves multilpe dynamic labels and finds the underlying metric.
+    ///
+    /// Expects a tuple of labels as `LS` argument.
+    pub fn resolve_labels<LS>(&self, labels: LS) -> &<Self as ResolveLabels<LS>>::Target
+    where
+        Self: ResolveLabels<LS>,
+    {
+        ResolveLabels::resolve_labels(self, labels)
+    }
+}
+
+/// Metric label using an `enum` as its value.
+///
+/// The most efficient way to specify metric labels in runtime, metric lookups
+/// using enum labels are as fast as array indexing. Prefer using this when all
+/// of your possible label values are known at the compile time.
+#[derive(Clone, Copy, Debug)]
+pub struct EnumLabel<const NAME: LabelName, T>(T);
+
+impl<const NAME: LabelName, T> EnumLabel<NAME, T> {
+    /// Creates a new [`EnumLabel`].
+    pub fn new(e: impl Into<T>) -> Self {
+        Self(e.into())
+    }
+
+    /// Convert this [`EnumLabel`] into the inner [`Enum`].
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+/// `enum` used as the value in [`EnumLabel`]s.
+///
+/// To implement this `trait` you also need to derive [`Ordinalize`] for your
+/// `enum`.
+/// SAFETY: DO NOT use custom discriminant values (eg. `enum MyEnum { MyVariant
+/// = -1 }`), this will lead to either:
+/// - `panic` in runtime (only for builds with `debug_assertions`)
+/// - incorrect label resolution
+pub trait Enum: Copy + Ordinalize<VariantType = i8> {
+    /// String representation of this enum.
+    fn as_str(&self) -> &'static str;
+}
+
+impl<const NAME: LabelName, T, M> DynamicLabel<M> for EnumLabel<NAME, T>
+where
+    T: Enum,
+{
+    // TODO: Switch to `[(T, M); T::VARIANT_COUNT]` once generic parameters are
+    // allowed to be used in const expressions.
+    type MetricCollection = Vec<(T, M)>;
```

Comment:
Since this PR is focused on micro-optimizations, we should consider stack-allocated `SmallVec` for this.

Reply:
It won't be on the stack anyway, as it's always `static`. And it will only allocate once per metric during the initialization
I guess it will remove a single indirection (heap pointer) for happy-path scenarios

Reply:
Yeah, the idea is to reduce indirection and allocations on push/resize.